helix-view/src/editor.rs

```diff
@@ -1187,6 +1189,87 @@ impl Editor {
         self.launch_language_servers(doc_id)
     }
 
+    /// moves/renames a path, invoking any event handlers (currently only lsp)
+    /// and calling `set_doc_path` if the file is open in the editor
+    pub fn move_path(&mut self, old_path: &Path, new_path: &Path) -> io::Result<()> {
+        let new_path = get_canonicalized_path(new_path);
+        // sanity check
+        if old_path == new_path {
+            return Ok(());
+        }
+        let is_dir = old_path.is_dir();
+        let language_servers: Vec<_> = self
+            .language_servers
+            .iter_clients()
+            .filter(|client| client.is_initialized())
+            .cloned()
+            .collect();
+        for language_server in language_servers {
+            let Some(request) = language_server.will_rename(old_path, &new_path, is_dir) else {
+                continue;
+            };
+            let edit = match helix_lsp::block_on(request) {
+                Ok(edit) => edit,
+                Err(err) => {
+                    log::error!("invalid willRename response: {err:?}");
+                    continue;
+                }
+            };
+            if let Err(err) = self.apply_workspace_edit(language_server.offset_encoding(), &edit) {
+                log::error!("failed to apply workspace edit: {err:?}")
+            }
+        }
+        fs::rename(old_path, &new_path)?;
```

Comment:
we could do that but it becomes pretty hard to do. The LSP standard theoretically allows the client to specify failure modes. The strategy helix (and VSCode) implement essentially just aborts on the first error. There is also a transactional rollback strategy but since VSCode doesn't support it basically no LS actually handles that so its not really practical to implement (altough its unclear to me why the failure modes are even necessary since all the state/transactions are managed by the client anyway). It may be possible to implement in the future but it would apply to way more code and would be really large change (and somewhat novel feature). I wouldn't want to tackle this in this PR